maven-surefire-common/src/test/java/org/apache/maven/plugin/surefire/util/DependenciesScannerTest.java

```diff
@@ -69,19 +78,141 @@ public void testLocateTestClasses()
+                                          "org/test/other/TestAA.class", "org/test/TestB.class" );
+        Artifact testArtifact =
+            new DefaultArtifact( "org.surefire.dependency", "dependent-artifact2",
+                                 VersionRange.createFromVersion( "1.0" ), "test", "jar", "tests-jdk15", null );
+        testArtifact.setFile( testJarFile );
+
+        List<String> scanDependencies = new ArrayList<String>();
+        scanDependencies.add( "org.surefire.dependency:dependent-artifact2:::tests-jdk15" );
+
+        List<String> include = new ArrayList<String>();
+        include.add( "**/*A.java" );
+        List<String> exclude = new ArrayList<String>();
+
+        DependencyScanner scanner =
+            new DependencyScanner( DependencyScanner.filter( Collections.singletonList( testArtifact ),
+                                                             scanDependencies ),
+                                   new TestListResolver( include, exclude ) );
+
+        ScanResult classNames = scanner.scan();
+        assertNotNull( classNames );
+        assertEquals( 2, classNames.size() );
+
+        Map<String, String> props = new HashMap<String, String>();
+        classNames.writeTo( props );
+        assertEquals( 2, props.size() );
+    }
+
+    /**
+     * Test with type when two artifacts are present, should only find the class in jar with correct type
+     */
+    @Test
+    public void testLocateTestClassesFromMultipleArtifactsWithType()
+        throws Exception
+    {
+        File jarFile =
+            writeTestFile( "DependenciesScannerTest3-1.0.jar", "org/test/ClassA.class", "org/test/ClassB.class" );
+        Artifact mainArtifact = new DefaultArtifact( "org.surefire.dependency", "dependent-artifact3",
+                                                     VersionRange.createFromVersion( "1.0" ), "test", "jar", null,
+                                                     new DefaultArtifactHandler() );
+        mainArtifact.setFile( jarFile );
+
+        File testJarFile =
+            writeTestFile( "DependenciesScannerTest3-1.0-tests.jar", "org/test/TestA.class", "org/test/TestB.class" );
+        Artifact testArtifact = new DefaultArtifact( "org.surefire.dependency", "dependent-artifact3",
+                                                     VersionRange.createFromVersion( "1.0" ), "test", "test-jar", null,
+                                                     new DefaultArtifactHandler() );
+        testArtifact.setFile( testJarFile );
+
+        List<String> scanDependencies = new ArrayList<String>();
+        scanDependencies.add( "org.surefire.dependency:dependent-artifact3::test-jar" );
+
+        List<String> include = new ArrayList<String>();
+        include.add( "**/*A.java" );
+        List<String> exclude = new ArrayList<String>();
+
+        DependencyScanner scanner =
+            new DependencyScanner( DependencyScanner.filter( Arrays.asList( mainArtifact, testArtifact ),
```

Comment:
Try to make the constructor call more compact using e.g. static imports or extracting to another local variables.